composer/trainer/trainer.py

```diff
@@ -465,6 +465,16 @@ def _generate_run_name() -> str:
     return generated_run_name
 
 
+def _get_sampler(dataloader: DataLoader) -> Sampler | Iterable:
+    """Fetch the sampler from a `dataloader`.
+
+    Returns `dalaoder.batch_sampler` is defined, else `dataloader.sampler` (always defined in `Dataloader.__init__`).
+    """
+    if dataloader.batch_sampler is not None:
+        return dataloader.batch_sampler
+    return dataloader.sampler
```

Comment:
Are the two types of samplers mutually exclusive?

Reply:
@antoinebrl  While instantiating yes, `sampler` and `batch_sampler` are mutually exclusive.
However, in `torch.utils.data.DataLoader.__init__`, `self.sampler` is instantiated with default sampler is None is provided, regardless of `batch_sampler` value (see [here](https://github.com/pytorch/pytorch/blob/23a6d74f9352e0afb37750fee300d077c4ba9393/torch/utils/data/dataloader.py#L344)). It will not be used if you provided a `batch_sampler`.

So the process if first to check is `batch_sampler` is available, else return the regular `sampler`.